Massimiliano Scarpetta

MaxScarp opened the pull request.

MaxScarp committed:
Development

Created and implemented first few functions for drawing lines.

MaxScarp committed:
Merge branch 'master' into pt-massimilianoScarpetta

# Conflicts:
#	appl/include/color.h
#	appl/include/line_raster.h
#	appl/include/scene.h
#	appl/include/screen.h
#	appl/src/line_raster.c
#	appl/src/scene.c
#	appl/src/screen.c

MaxScarp committed:
Retrieve all info but faces from obj file

TODO - creating a struct for reading and saving faces info

MaxScarp committed:
Fix, refactoring and created a function to make vertices

TODO - function to make triangles

MaxScarp committed:
Fixed a bug on obj_new and obj_free + added all comments

MaxScarp committed:
Finished obj_parser refactoring and added stormtrooper test

MaxScarp committed:
Updated obj_parser function

Now the function can save inside the obj structure also other counters and not only triangle amount

fdefelici reviewed (changes requested):
Good you tried to avoid the double while cycle and the attempt to improve code quality by splitting behavior in proper function.

Possibly for making obj_parser a real single-header library, try to add an OBJ_IMPLEMENTATION macro to separate declarations from definitions (like stb_image does or as we saw in clove-unit)

Here some hint about the implementation:
- you avoided double while, but realloc at each `obj_add_position` is even worse in terms of performance :). Try at least to realloc at block of memory (each 10, 100, ....)
- give a signal to a possible user to not use internal apis (like `obj_add_position`). Prepend an `_` before the function name could help (like in clove-unit for instance)
- Separate tests at least in two file. One dedicate for main apis `obj_parse` and `obj_free` and another one to test internal apis (like the ones for the array)
- in `obj_t` rename `triangle_dynamic_array` to just `triangles` (the end user does not need to know it is a dynamic array :) )

Futhermore you could make few codes improvements:
- replace `rewind` with more secure `fseek` call
- replace `atoi` and `atof` with more secure `strtol and `strtof`

MaxScarp committed:
Code refactoring

- Now the dynamic arrays are pre-allocated with a capacity of 10 and decuplicate their capacity once the capacity has been reaced.
- All the functions that the user should not use have been renamed using a "_" before the function name.
- Separated test into two files (obj_parset_test, obj_api_test).
- Renamend triangle_dynamic_array inside obj_t into triangles.
- Replaced usage of atoi with strtol and usage of atof with strtof.

MaxScarp committed:
Finished refactoring

Added one more "_" before the function names that the user should not use. Added the new memory allocation management also for "triangles" and "vertex_dynamic_array"

fdefelici commented on appl/include/obj_parser.h:
```diff
@@ -1,181 +1,553 @@
+/// @param vertex2 The second vertex od the triangle
+/// @param vertex3 The third vertex od the triangle
+/// @return A new triangle structure
+obj_triangle_t __obj_triangle_new(obj_vertex_t vertex1, obj_vertex_t vertex2, obj_vertex_t vertex3)
```
this could be tricky in termes of data copying. The compiler could be smart enough to inline this function, but in the worse case this would happen:
- Each paramters is copied when function is invoked
- Then another copy of the parameters happen when assigning to v1, v2, v3
- Then a temporary result triangle is crated that will be copied in the array.

Looking at usage of this function (just one call), maybe could be better to inline the code where it is needed.

MaxScarp replied:
Are you suggesting to do something like this? return (obj_triangle_t){vertex1, vertex2, vertex3;

fdefelici replied:
nope. Basically remove that method and just use the internal code in the only one point where is needed

fdefelici commented on appl/include/obj_parser.h:
```diff
@@ -1,181 +1,553 @@
+/// @param obj A pointer to the obj structure
+/// @param triangle The triangle structure to be added
+/// @return An error number, 0 if everything have been completed succesfully
+int __obj_add_triangle(obj_t* obj, obj_triangle_t triangle)
```
here the whole triangle is copied when passed as parameters, and then copied again when doing the assignement.

In this case, could be better to pass the triangle as pointer and then deferencing it when assigning reducing to 1 copy.

fdefelici commented on test/src/obj_parser_test.c:
```diff
@@ -1,90 +1,265 @@
 #define CLOVE_SUITE_NAME ObjParserTest
+#define OBJ_IMPLEMENTATION
```
About OBJ_IMPLEMENTATION macro in order to work properly you have to re-organize the obj_parser.h in order to split definition and implementations:

```
//Definitions
struct hello_t { };

void hello(struct hello_t* );

//Implementations
#ifdef HELLO_IMPLEMENTATON

void hello(struct hello_t* )
{
   //hello function implementation
}

#endif



```

MaxScarp committed:
Code refactoring

- Removed __obj_triangle_new for better performance
- Modified __obj_add_triangle so that it takes a pointer to an obj_triangle_t structure instead of make a copy of the structure
- Modified preprocessor definition to make everything more similar to a single header library

fdefelici commented on appl/include/obj_parser.h:
```diff
@@ -1,181 +1,570 @@
 
-            for (int i = 0; i < 3; i++)
+    char buffer[1024];
+    while(fgets(buffer, sizeof(buffer), file))
```
the 4 `if` checking for a tag (v, vt, vn and f), are always executed (even if only one is validated to true)

Improve performance using 1 if and 3 else if.

fdefelici commented on appl/include/obj_parser.h:
```diff
@@ -1,181 +1,570 @@
+
+/// @brief Function to free the memory allocated for an obj
+/// @param obj The obj structure to free
+void __obj_free(obj_t* obj);
```
obj_free must be available for users in order to destroy an obj created with obj_parse function.

MaxScarp replied:
Ops...

MaxScarp committed:
Code refactoring

Now obj_new and obj_free are available again for users

MaxScarp committed:
Code refactoring

- Added const to some function's argument
- The 4 if statements inside obj_parser have been replaced with 1 if and 3 else if

fdefelici commented on appl/include/obj_parser.h:
```diff
@@ -1,181 +1,567 @@
+{
+    obj_triangle_t* triangles;
+    size_t triangle_amount;
+    size_t triangles_capacity;
```
as far the obj_t is the "object" that a final user obtain, the `triangles_capacity` is not an information that need to be exposed to the final user. So better to avoid to store it in the `obj_t`

fdefelici replied:
as far the obj_t is the "object" that a final user obtain, the `triangles_capacity` is not an information that need to be exposed to the final user. So better to avoid to store it in the `obj_t`

fdefelici commented on appl/include/obj_parser.h:
```diff
@@ -1,181 +1,567 @@
+
+/// @brief Function to create a new obj and initialize its members
+/// @return A pointer to the obj structure
+obj_t* obj_new();
```
only 'obj_free' should be `public`. (it is necessary to the final user to delete properly an `obj_t` obtained using `obj_parse(...)` function

The final user never need to call `obj_new()` so should be kept as an internal with the double undescore

MaxScarp committed:
Code refactoring

- The __obj_new() function is no more "visible" to the final user
- The triangles_capacity have been moved from obj_t structure to obj_info_t structure
